cider/exec/nextgen/jitlib/llvmjit/LLVMJITModule.cpp

```diff
@@ -0,0 +1,93 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "exec/nextgen/jitlib/llvmjit/LLVMJITModule.h"
+
+#include <llvm/IR/Function.h>
+
+#include "exec/nextgen/jitlib/llvmjit/LLVMJITUtils.h"
+#include "util/Logger.h"
+
+namespace jitlib {
+
+LLVMJITModule::LLVMJITModule(const std::string& name)
+    : context_(std::make_unique<llvm::LLVMContext>())
+    , module_(std::make_unique<llvm::Module>(name, *context_))
+    , engine_(nullptr) {
+  CHECK(context_);
+  CHECK(module_);
+}
+
+static llvm::FunctionType* getFunctionSignature(const JITFunctionDescriptor& descriptor,
+                                                llvm::LLVMContext& context) {
+  llvm::Type* ret_type = getLLVMType(descriptor.ret_type.type, context);
+
+  llvm::SmallVector<llvm::Type*, 8> arguments;
+  for (const JITFunctionParam& param_descriptor : descriptor.params_type) {
+    llvm::Type* arg_type = getLLVMType(param_descriptor.type, context);
+    if (arg_type) {
+      arguments.push_back(arg_type);
+    } else {
+      // TODO (bigPYJ1151): Add Exceptions.
+    }
+  }
+
+  return llvm::FunctionType::get(ret_type, arguments, false);
+}
+
+LLVMJITFunction LLVMJITModule::createJITFunctionImpl(
+    const JITFunctionDescriptor& descriptor) {
+  auto func_signature = getFunctionSignature(descriptor, *context_);
+  llvm::Function* func = llvm::Function::Create(func_signature,
+                                                llvm::GlobalValue::ExternalLinkage,
+                                                descriptor.function_name,
+                                                *module_);
+  func->setCallingConv(llvm::CallingConv::C);
+
+  auto arg_iter = func->arg_begin();
+  for (size_t index = 0; index < descriptor.params_type.size(); ++index, ++arg_iter) {
+    if (auto name = descriptor.params_type[index].name; name) {
+      arg_iter->setName(name);
+    }
+  }
+
+  // TODO (bigPYJ1151): Set Parameters Attributes.
+  return LLVMJITFunction(descriptor, *this, *func);
+}
+
+void LLVMJITModule::finishImpl() {
+  // TODO (bigPYJ1151): Refactor Debug information.
+  llvm::outs() << *module_;
+
+  LLVMJITEngineBuilder builder(*this);
+  engine_ = builder.build();
+}
+
+void* LLVMJITModule::getFunctionPtrImpl(LLVMJITFunction& function) {
```

Comment:
Q; why not using function descriptor directly? Instead of looking up by ```LLVMJITFunction```. My understanding, we will maintain a codegen context hosting all functions outside JITModule. If so, we can directly use ```LLVMJITFunction```. 

Your thoughts?

Reply:
This function is used for function pointer lookup and a function pointer executable. If we want to build a Call instruction, just provide function name or LLVMJITFunction and needn't to call getFunctionPtrImpl.

It's feasible to move this function as a member function of LLVMJITFunction because it has a reference to the JITModule.